### ISSUE TITLE
Create configuration options for EF Core database tracing

### ISSUE THREAD

**Issue by jderose9** [elastic/apm-agent-dotnet#681]
**Is your feature request related to a problem? Please describe.**
There is no way to disable the EF Core database tracing, or apply limits to the amount of data captured, by way of configuration. (In contrast to stack trace and HTTP Request tracing, for example)

**Describe the solution you'd like**
Allow us to enable/disable database tracing via config, and to set limits for how many database spans specifically are captured and/or a time threshold for minimum run length before a span is captured (similar to stack trace capture limits).

**Describe alternatives you've considered**
Using another APM tool

**Comment by gregkalapos** [comment]
> There is no way to disable the EF Core database tracing

You can disable it - that is already something you can change when you configure the agent - [this part of the doc](https://www.elastic.co/guide/en/apm/agent/dotnet/current/setup.html#setup-asp-net-core) describes it. 

So if you activate the agent like this:

```
app.UseElasticApm(subscribers: new HttpDiagnosticsSubscriber());
```

then EF Core calls won't be captured - only outgoing HTTP calls will be captured. So disabling it is possible. `EfCoreDiagnosticsSubscriber` is the one that captures db and that is automatically added if you use the `UseAllElasticApm()` method (more details in the linked docs).

To the the other parts:

There is a new feature in the making that could help here I feel - [this PR](https://github.com/elastic/apm-agent-dotnet/pull/792) will introduce filters and with that you'll be able to control the data that is sent to the APM Server.

E.g.:

```
Agent.AddFilter((ISpan span) =>
{
	// If it captured DB info
	if (span?.Context?.Db != null)
	{
		// and the duration is less than x
		if (span.Duration < 500)
		{
			//Do not send any db related info
			span.Context.Db = null;
		}
	}
	return true;
});
```

Of course this way the agent will still capture the db info, but you can omit it before it'd be sent. From the performance point of view I'm not sure we'd save anything with a config - since the duration is only known after the db call is executed we'd need to create the span when the db call is started anyway - so i'd estimate that it does not make a difference if you throw it away in a filter vs. the agent throws it away once the db call is done. 

The way stacktrace capturing works is that once a span is finished the agent will either capture the stacktrace or not based on a duration threshold - but capturing db calls is different - we need to create the span once the db call starts, but part of the data will be only available once the db call finished. Also limiting the number of db calls on let's say a transaction is not that trivial - this would mean that we need to traverse all the spans on a transaction each time a db call happens. So overall I feel that from the perf. point of view filters should be acceptable.

Does this cover your use case?

**Comment by jderose9** [comment]
Hi @gregkalapos! Yes, I am aware of the option to pass subscribers in (the documentation was added per [another issue of mine](https://github.com/elastic/apm-agent-dotnet/issues/680)). There is no way to do this via configuration options as I can in the other modules, though. I could implement my own configuration driven logic - but I guess this is more of a quality-of-life request, in that it is confusing that there is not consistency between the modules.

Regarding the duration settings - I essentially have many small database requests which are getting captured during the course of a transaction and I largely am not concerned about them if they aren't misbehaving. The fill up the transaction timeline with a lot of data, and their span names are the full query text which may be quite long. So I'm basically transmitting a lot of data which then clogs up the ability to quickly diagnose a problem. It does look like the filter PR you are describing would allow me to write custom logic to handle this (yay!), but again - it's available in the stack trace module and having consistency would make implementation much faster for new users.

**Comment by gregkalapos** [comment]
Thanks for the feedback @jderose9.

Maybe one thing to keep in mind here is that EF Core tracing is very different from stack trace capturing. Again, the `EfCoreDiagnosticsSubscriber` creates spans with database information - what the [Stacktrace config options](https://www.elastic.co/guide/en/apm/agent/dotnet/current/config-stacktrace.html) do is that they simply control if and how a stacktrace should be captured on an existing span - it does not control the span creation itself. So I’m not sure there should be any consistency between these two things - what you propose would be about whether to create a span or not, the stacktrace config is about whether to capture additional context (callstack) on an existing span. Also, the code that captures stacktraces is part of the core agent, it’s always present if the agent is active unlike `EfCoreDiagnosticsSubscriber`.

Honestly I’m still a bit reluctant to add these configs from the following reasons:

1.  Regarding a setting to turn off EF Core tracing: currently there is no such setting for any “module”.
You say: "There is no way to do this via configuration options as I can in the other modules, though."
What specific setting do you refer to here? In case you mean the stack trace setting (e.g. setting `StackTraceLimit` to 0 to turn off stacktrace collection is different) then that still does not control whether a span is created or not, it’s only about capturing stacktrace for existing spans and that setting can be used in every possible setup of the agent).
The issue here is that EF Core capturing is not always present - e.g. you can have a full framework application without EF Core, and in this case this setting would not do anything, because the code that’d read the setting is simply not present in the given configuration. I feel overall this would be more confusing and open up more questions. The stacktrace settings are different - those are part of the core agent.

2. Regarding adding settings similar to stack trace capturing: as already partly said, I got your point about consistency, but I’m not sure from other users point of view this would be consistent. Capturing DB calls based on a duration would mean that a span is either created or not - so you’ll either not see anything from the db call or you’ll see it. Capturing callstack is only about the context of the given span, the span itself will always exist. Plus in this case we also need to keep in mind that in some setups in case `EfCoreDiagnosticsSubscriber` is not present such a setting would not do anything.

On the other hand maybe some general db span creation setting would be useful - and then whatever module creates spans with db information, then those could read those configs. But from what I see, at the moment no other agent has such configs, and so far I haven’t seen other requests for such configs.

**Comment by jderose9** [comment]
Thanks @gregkalapos. Just to clarify, when I have been saying "capture", I am referring to the data sent to and captured by the elasticsearch server, rather than the agent. Ultimately, I just want as much as control as possible to tune the amount of data that I am capturing during a trace.

I understand your point about it not being a perfect analog to the other configuration sections because this would affect span creation. I do think they are similar in that they are pieces of the trace that can possibly contain a lot of data, except there is no way to tune the amount of data that is sent. Your concern about missing spans makes sense, but ultimately I would not personally be bothered by that. In any trace there is a lot of activity happening between the spans, or within the spans, that is just not captured with the auto-instrumentation - File IO, internal calls, etc.

Coming in as a new user, I examined the Configuration section of the documentation:
![image](https://user-images.githubusercontent.com/2118702/78416150-d0933600-75f4-11ea-8a65-ad7d20459029.png)

In the HTTP Configuration options, I get several choices that can significantly affect the amount of data that ends up in the trace, including setting CaptureBody to off. I can even set this from the central config, without touching the application or device it runs on.

In the Stacktrace Configuration options, I again get several choices to limit the data shipped to the server, including just setting to 0 to turn it off.

In the Entity Framework Configuration options - well, that section doesn't exist - there aren't any. I can't do anything but ship every single query that executes during the trace to the server, and to turn it off I need to make a compiled code change or write some custom logic to handle it.

Ultimately I am not too bothered either way whether you decide to implement those things, so feel free to close this - I just wanted to share what I thought was a pain point for me as a new user.

**Comment by gregkalapos** [comment]
Great, thank you for the feedback @jderose9. 

Yeah, I'm still thinking about adding maybe these settings as "database" settings. Once we get further requests to this we can definitely revisit.

Closing for now.
